src/OmniSharp.MiscellaneousFile/MiscellaneousFileProjectSystem.cs

```diff
@@ -0,0 +1,116 @@
+        public string Key => ProjectSystemNames.MiscellaneousFilesProjectSystem;
+        public string Language => LanguageNames.CSharp;
+        IEnumerable<string> IProjectSystem.Extensions => new[] { miscFileExtension };
+        public bool EnabledByDefault { get; } = true;
+
+        private readonly ConcurrentDictionary<string, DocumentId> _documents = new ConcurrentDictionary<string, DocumentId>();
+        private readonly OmniSharpWorkspace _workspace;
+        private readonly IFileSystemWatcher _fileSystemWatcher;
+        private readonly FileSystemHelper _fileSystemHelper;
+        private readonly List<IWaitableProjectSystem> _projectSystems;
+        private readonly ILogger _logger;
+
+        [ImportingConstructor]
+        public MiscellaneousFilesProjectSystem(OmniSharpWorkspace workspace, IFileSystemWatcher fileSystemWatcher, FileSystemHelper fileSystemHelper,
+            ILoggerFactory loggerFactory, [ImportMany] IEnumerable<Lazy<IProjectSystem, ProjectSystemMetadata>> projectSystems)
+        {
+            _workspace = workspace;
+            _fileSystemWatcher = fileSystemWatcher;
+            _fileSystemHelper = fileSystemHelper;
+            _logger = loggerFactory.CreateLogger<MiscellaneousFilesProjectSystem>();
+            _projectSystems = projectSystems
+                .Where(ps => ps.Metadata.Name == ProjectSystemNames.MSBuildProjectSystem ||
+                 ps.Metadata.Name == ProjectSystemNames.DotNetProjectSystem)
+                .Select(ps => ps.Value)
+                .Cast<IWaitableProjectSystem>()
+                .ToList();
+        }
+
+        Task<object> IProjectSystem.GetProjectModelAsync(string filePath)
+        {
+            return Task.FromResult<object>(null);
+        }
+
+        Task<object> IProjectSystem.GetWorkspaceModelAsync(WorkspaceInformationRequest request)
+        {
+            return Task.FromResult<object>(null);
+        }
+
+        void IProjectSystem.Initalize(IConfiguration configuration)
+        {
+            var allFiles = _fileSystemHelper.GetFiles("**/*.cs");
+            foreach (var filePath in allFiles)
+                AddIfMiscellaneousFile(filePath);
+
+            _fileSystemWatcher.Watch(miscFileExtension, OnMiscellaneousFileChanged);
+        }
+
+        private async void AddIfMiscellaneousFile(string filePath)
+        {
+            //wait for the project systems to finish processing the updates
+            foreach (var projectSystem in _projectSystems)
+            {
+                await projectSystem.WaitForUpdatesAsync();
+            }
+
+            var absoluteFilePath = new FileInfo(filePath).FullName;
+            if (!File.Exists(absoluteFilePath))
+                return;
+
+            var documentId = _workspace.TryAddMiscellaneousFileDocument(absoluteFilePath, Language);
+            if (documentId!= null)
+            {
+                _documents[absoluteFilePath] = documentId;
```

Comment:
Do we still need to track these documents in the project system? Seems like the workspace now knows how to do the right thing?